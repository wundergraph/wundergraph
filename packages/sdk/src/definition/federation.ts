import { GraphQLSchema } from 'graphql';
import axios from 'axios';

export const isFederationService = (schema: GraphQLSchema): boolean => {
	const queryType = schema.getQueryType();
	if (queryType === undefined || queryType === null) {
		return false;
	}
	const fields = queryType.getFields();
	if (fields === undefined) {
		return false;
	}
	return Object.keys(fields).indexOf('_service') !== -1;
};

export const fetchFederationServiceSDL = async (url: string): Promise<string> => {
	const data = JSON.stringify({
		query: '{_service{sdl}}',
	});
	const res = await axios.post(url, data, {
		headers: {
			'Content-Type': 'application/json',
			Accept: 'application/json',
		},
	});
	return res.data.data._service.sdl;
};
