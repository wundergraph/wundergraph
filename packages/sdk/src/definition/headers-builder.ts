import { EnvironmentVariable, InputVariable, PlaceHolder } from '../configure/variables';
import { ConfigurationVariable, ConfigurationVariableKind, HTTPHeader } from '@wundergraph/protobuf';

export interface HeaderConfiguration {
	// key is the name of the Header that will be sent to the upstream, e.g. Authorization
	key: string;
	// valueSource defines where the value should come from
	// static means, use the static content of the field "value"
	// clientRequest means, the content of "value" defines a client request Header field
	// so, if the field value is "Authorization",
	// the client Header value for the key "Authorization" will be injected into the header
	// env means, use the value from the environment variable
	valueSource: 'static' | 'env' | 'clientRequest' | 'placeholder';
	value: string;
	defaultValue?: string;
}

export interface IGraphqlIntrospectionHeadersBuilder {
	addStaticHeader: (key: string, value: InputVariable) => IGraphqlIntrospectionHeadersBuilder;
}

export interface IHeadersBuilder {
	addStaticHeader: (key: string, value: InputVariable) => IHeadersBuilder;
	addClientRequestHeader: (key: string, requestHeaderName: string) => IHeadersBuilder;
}

export class HeadersBuilder {
	private headers: HeaderConfiguration[] = [];

	public addStaticHeader = (key: string, value: InputVariable) => {
		if (value === undefined) {
			throw new Error(`Static header value cannot be undefined`);
		}
		if (typeof value === 'string') {
			this.headers.push({
				key,
				value,
				valueSource: 'static',
			});
			return this;
		}
		if ((value as PlaceHolder)._identifier === 'placeholder') {
			this.headers.push({
				key,
				value: (value as PlaceHolder).name,
				valueSource: 'placeholder',
			});
			return this;
		}
		this.headers.push({
			key,
			value: value.name,
			valueSource: 'env',
			defaultValue: (value as EnvironmentVariable).defaultValue,
		});
		return this;
	};
	public addClientRequestHeader = (key: string, requestHeaderName: string) => {
		this.headers.push({
			key,
			value: requestHeaderName,
			valueSource: 'clientRequest',
		});
		return this;
	};
	public build = (): HeaderConfiguration[] => {
		return this.headers;
	};
}

export const mapHeaders = (builder: HeadersBuilder): { [key: string]: HTTPHeader } => {
	const headers: { [key: string]: HTTPHeader } = {};

	builder.build().forEach((config: HeaderConfiguration) => {
		const values: ConfigurationVariable[] = [];
		switch (config.valueSource) {
			case 'placeholder':
				values.push({
					kind: ConfigurationVariableKind.PLACEHOLDER_CONFIGURATION_VARIABLE,
					staticVariableContent: '',
					environmentVariableDefaultValue: '',
					environmentVariableName: '',
					placeholderVariableName: config.value,
				});
				break;
			case 'clientRequest':
				values.push({
					kind: ConfigurationVariableKind.STATIC_CONFIGURATION_VARIABLE,
					staticVariableContent: `{{ .request.headers.${config.value} }}`,
					environmentVariableName: '',
					environmentVariableDefaultValue: '',
					placeholderVariableName: '',
				});
				break;
			case 'static':
				values.push({
					kind: ConfigurationVariableKind.STATIC_CONFIGURATION_VARIABLE,
					staticVariableContent: config.value,
					environmentVariableDefaultValue: '',
					environmentVariableName: '',
					placeholderVariableName: '',
				});
				break;
			case 'env':
				values.push({
					kind: ConfigurationVariableKind.ENV_CONFIGURATION_VARIABLE,
					staticVariableContent: '',
					environmentVariableDefaultValue: config.defaultValue || '',
					environmentVariableName: config.value,
					placeholderVariableName: '',
				});
				break;
		}
		headers[config.key] = {
			values,
		};
	});

	return headers;
};
