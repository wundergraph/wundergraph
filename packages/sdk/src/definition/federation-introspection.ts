import { DocumentNode, parse } from 'graphql';
import { AxiosError, AxiosRequestConfig, AxiosResponse } from 'axios';
import { GraphQLApi, GraphQLFederationIntrospection, GraphQLIntrospection, ApiIntrospectionOptions } from './index';
import { loadFile } from '../codegen/templates/typescript';
import { resolveVariable } from '../configure/variables';
import { mergeApis } from './merge';
import { introspectWithCache } from './introspection-cache';
import {
	graphqlIntrospectionCacheConfiguration,
	introspectGraphql,
	resolveGraphqlIntrospectionHeaders,
} from './graphql-introspection';
import { HeadersBuilder, mapHeaders } from './headers-builder';
import { Fetcher } from './introspection-fetcher';
import { Logger } from '../logger';

export const fetchFederationServiceSDL = async (
	url: string,
	headers?: Record<string, string>,
	meta?: { apiNamespace?: string; upstreamName?: string }
): Promise<string> => {
	const data = JSON.stringify({
		query: '{_service{sdl}}',
	});

	let opts: AxiosRequestConfig = {
		headers,
		'axios-retry': {
			onRetry: (retryCount: number, error: AxiosError, requestConfig: AxiosRequestConfig) => {
				let msg = `failed to fetch federation service sdl: ${requestConfig.method} url: ${requestConfig.url}`;
				if (meta?.apiNamespace) {
					msg += ` apiNamespace: ${meta.apiNamespace}`;
				}
				if (meta?.upstreamName) {
					msg += ` upstreamName: ${meta.upstreamName}`;
				}

				msg += ` retryAttempt: ${retryCount}`;

				Logger.info(msg);
			},
		},
	};

	let res: AxiosResponse | undefined;
	try {
		res = await Fetcher().post(url, data, opts);
	} catch (e: any) {
		throw new Error(`failed to fetch federation service sdl (url: ${url}), error: ${e.message}`);
	}
	if (res === undefined) {
		throw new Error(`failed to fetch federation service sdl (url: ${url}), no response`);
	}
	if (res.status !== 200) {
		throw new Error(
			`failed to fetch federation service sdl (url: ${url}), response code: ${res.status}, message: ${res.statusText}`
		);
	}

	return res.data.data._service.sdl;
};

export interface ServiceDefinition {
	typeDefs: DocumentNode;
	name: string;
}

export const introspectFederation = async (introspection: GraphQLFederationIntrospection) => {
	const keyInputs: string[] = [];
	introspection.upstreams.forEach(async (upstream) => {
		keyInputs.push(await (await graphqlIntrospectionCacheConfiguration(upstream)).keyInput);
	});
	const cacheConfig = { keyInput: keyInputs.join('') };
	return introspectWithCache(
		introspection,
		cacheConfig,
		async (introspection: GraphQLFederationIntrospection, options: ApiIntrospectionOptions): Promise<GraphQLApi> => {
			const upstreams = introspection.upstreams.map(async (upstream, i) => {
				let schema = upstream.loadSchemaFromString ? loadFile(upstream.loadSchemaFromString) : '';

				const name = upstream.name ?? i.toString();

				if (schema === '' && upstream.url) {
					const introspectionHeadersBuilder = new HeadersBuilder();

					if (upstream.headers !== undefined) {
						upstream.headers(introspectionHeadersBuilder);
					}

					if (upstream.introspection?.headers !== undefined) {
						upstream.introspection?.headers(introspectionHeadersBuilder);
					}
					const introspectionHeaders = resolveGraphqlIntrospectionHeaders(mapHeaders(introspectionHeadersBuilder));

					// upstream.url is truthy at this point, no need to check
					schema = await fetchFederationServiceSDL(resolveVariable(upstream.url), introspectionHeaders, {
						apiNamespace: introspection.apiNamespace,
						upstreamName: name,
					});
				}

				if (schema == '') {
					throw new Error(`Subgraph ${name} has not provided a schema`);
				}

				return {
					name,
					typeDefs: parse(schema),
				};
			});

			const graphQLIntrospections: GraphQLIntrospection[] = introspection.upstreams.map((upstream) => ({
				...upstream,
				apiNamespace: introspection.apiNamespace,
				id: upstream.id ?? introspection.id,
				isFederation: true,
			}));

			const apis = await Promise.all(graphQLIntrospections.map((i) => introspectGraphql(i, options)));
			const merged = mergeApis([], [], ...apis) as GraphQLApi;
			merged.Namespace = introspection.apiNamespace || '';
			return merged;
		}
	);
};
