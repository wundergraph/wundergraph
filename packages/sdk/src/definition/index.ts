import { buildClientSchema, GraphQLSchema, introspectionFromSchema, parse, print, printSchema } from 'graphql';
import { renameTypeFields, renameTypes } from '../graphql/renametypes';
import {
	ArgumentSource,
	ConfigurationVariable,
	DataSourceKind,
	DirectiveConfiguration,
	FetchConfiguration,
	FieldConfiguration,
	GraphQLDataSourceHooksConfiguration,
	MTLSConfiguration,
	SigningMethod,
	SingleTypeField,
	StatusCodeTypeMapping,
	TypeConfiguration,
	TypeField,
	UpstreamAuthentication,
	UpstreamAuthenticationKind,
} from '@wundergraph/protobuf';
import { applyNameSpaceToGraphQLSchema } from './namespacing';
import { InputVariable, mapInputVariable } from '../configure/variables';
import { introspectGraphqlWithCache } from './graphql-introspection';
import { introspectFederation } from './federation-introspection';
import { IGraphqlIntrospectionHeadersBuilder, IHeadersBuilder } from './headers-builder';
import { introspectOpenApi, introspectOpenApiV2 } from './openapi-introspection';
import {
	introspectMongoDB,
	introspectMySQL,
	introspectPlanetScale,
	introspectPostgresql,
	introspectPrisma,
	introspectSQLite,
	introspectSQLServer,
} from './database-introspection';
import { introspectSoap } from './soap-introspection';

// Use UPPERCASE for environment variables
export const WG_DATA_SOURCE_POLLING_MODE = process.env['WG_DATA_SOURCE_POLLING_MODE'] === 'true';
export const WG_ENABLE_INTROSPECTION_CACHE = process.env['WG_ENABLE_INTROSPECTION_CACHE'] === 'true';
export const WG_INTROSPECTION_CACHE_SKIP = process.env['WG_INTROSPECTION_CACHE_SKIP'] === 'true';
// Only use the introspection cache, return an error when hitting the network
export const WG_ENABLE_INTROSPECTION_OFFLINE = process.env['WG_ENABLE_INTROSPECTION_OFFLINE'] === 'true';

export const WG_PRETTY_GRAPHQL_VALIDATION_ERRORS = process.env['WG_PRETTY_GRAPHQL_VALIDATION_ERRORS'] === 'true';
// Default polling interval for data sources without one
export const WG_DATA_SOURCE_DEFAULT_POLLING_INTERVAL_SECONDS = (() => {
	const seconds = parseInt(process.env['WG_DATA_SOURCE_DEFAULT_POLLING_INTERVAL_SECONDS'] ?? '', 10);
	return isNaN(seconds) ? 0 : seconds;
})();

/**
 * ApiIntrospectionOptions contains options that are passed to
 * all ApiIntrospector<T> functions
 */
export interface ApiIntrospectionOptions {
	/**
	 * Global proxy URL, which might be overridden at the data source level
	 */
	httpProxyUrl?: string;
	apiID?: string;
}

export interface RenameType {
	from: string;
	to: string;
}

export interface RenameTypes {
	renameTypes: (rename: RenameType[]) => void;
}

export interface RenameTypeField {
	typeName: string;
	fromFieldName: string;
	toFieldName: string;
}

export interface RenameTypeFields {
	renameTypeFields: (rename: RenameTypeField[]) => void;
}

export type ApiType = GraphQLApiCustom | RESTApiCustom | DatabaseApiCustom;

export class Api<T = ApiType> implements RenameTypes, RenameTypeFields {
	constructor(
		schema: string,
		namespace: string,
		dataSources: DataSource<T>[],
		fields: FieldConfiguration[],
		types: TypeConfiguration[],
		interpolateVariableDefinitionAsJSON: string[],
		customJsonScalars?: string[]
	) {
		this.Schema = schema;
		this.Namespace = namespace;
		this.DataSources = dataSources;
		this.Fields = fields;
		this.Types = types;
		this.interpolateVariableDefinitionAsJSON = interpolateVariableDefinitionAsJSON;
		this.CustomJsonScalars = customJsonScalars;
	}

	DefaultFlushInterval: number = 500;
	Schema: string;
	DataSources: DataSource<T>[];
	Fields: FieldConfiguration[];
	Types: TypeConfiguration[];
	interpolateVariableDefinitionAsJSON: string[];
	CustomJsonScalars?: string[];
	Namespace: string;

	renameTypes(rename: RenameType[]): void {
		this.Schema = renameTypes(this.Schema, rename);
		this.DataSources = this.DataSources.map((d) => {
			return {
				...d,
				RootNodes: typeFieldsRenameType(d.RootNodes, rename),
				ChildNodes: typeFieldsRenameType(d.ChildNodes, rename),
			};
		});
		this.Fields = this.Fields.map((field) => {
			const re = rename.find((r) => r.from === field.typeName);
			return {
				...field,
				typeName: re !== undefined ? re.to : field.typeName,
			};
		});
	}

	renameTypeFields(rename: RenameTypeField[]): void {
		this.Schema = renameTypeFields(this.Schema, rename);
		this.DataSources = this.DataSources.map((d) => {
			return {
				...d,
				RootNodes: typeFieldsRenameTypeField(d.RootNodes, rename),
				ChildNodes: typeFieldsRenameTypeField(d.ChildNodes, rename),
			};
		});
		this.Fields = this.Fields.map((field) => {
			const re = rename.find((re) => re.typeName === field.typeName && re.fromFieldName === field.fieldName);
			if (re !== undefined) {
				return {
					...field,
					fieldName: re.toFieldName,
					path: field.path.map((item) => (item === field.fieldName ? re.toFieldName : item)),
				};
			}
			const sameTypeRenameFields = rename.filter((re) => re.typeName === field.typeName);
			return {
				...field,
				requiresFields: field.requiresFields.map((f) => {
					const re = sameTypeRenameFields.find((sameTypeField) => sameTypeField.fromFieldName === f);
					if (re !== undefined) {
						return re.toFieldName;
					}
					return f;
				}),
				argumentsConfiguration: field.argumentsConfiguration.map((arg) => {
					if (arg.sourceType === ArgumentSource.OBJECT_FIELD) {
						return {
							...arg,
							sourcePath: arg.sourcePath.map((item) => {
								const re = sameTypeRenameFields.find((sameTypeField) => sameTypeField.fromFieldName === item);
								if (re !== undefined) {
									return re.toFieldName;
								}
								return item;
							}),
						};
					}
					return arg;
				}),
			};
		});
	}
}

const typeFieldsRenameType = (fields: TypeField[], rename: RenameType[]): TypeField[] => {
	return fields.map((node) => {
		const re = rename.find((r) => r.from === node.typeName);
		return {
			...node,
			typeName: re !== undefined ? re.to : node.typeName,
		};
	});
};

const typeFieldsRenameTypeField = (fields: TypeField[], rename: RenameTypeField[]): TypeField[] => {
	return fields.map((node) => {
		return {
			...node,
			fieldNames: node.fieldNames.map((field) => {
				const re = rename.find((re) => re.typeName === node.typeName && re.fromFieldName === field);
				if (re !== undefined) {
					return re.toFieldName;
				}
				return field;
			}),
		};
	});
};

export const createMockApi = async (sdl: string, apiNamespace?: string): Promise<Api<any>> => {
	const schema = print(parse(sdl));
	return new GraphQLApi(applyNameSpaceToGraphQLSchema(schema, [], apiNamespace), apiNamespace || '', [], [], [], []);
};

export class GraphQLApi extends Api<GraphQLApiCustom> {}

export class RESTApi extends Api<RESTApiCustom> {}

export class PostgresqlApi extends Api<DatabaseApiCustom> {}

export class MySQLApi extends Api<DatabaseApiCustom> {}

export class PlanetscaleApi extends Api<DatabaseApiCustom> {}

export class SQLiteApi extends Api<DatabaseApiCustom> {}

export class SQLServerApi extends Api<DatabaseApiCustom> {}

export class MongoDBApi extends Api<DatabaseApiCustom> {}

export class PrismaApi extends Api<DatabaseApiCustom> {}

export interface DataSource<Custom = unknown> {
	Id?: string;
	Kind: DataSourceKind;
	RootNodes: TypeField[];
	ChildNodes: TypeField[];
	Custom: Custom;
	Directives: DirectiveConfiguration[];
	RequestTimeoutSeconds: number;
}

interface GraphQLIntrospectionOptions {
	// loadSchemaFromString allows you to skip the introspection process and load the GraphQL Schema from a string instead
	// this way, you can import a GraphQL Schema file or load the Schema in more flexible ways than relying on sending a GraphQL Introspection Query
	loadSchemaFromString?: string | (() => string);
	customFloatScalars?: string[];
	customIntScalars?: string[];
	customJSONScalars?: string[];
	// switching internal to true will mark the origin as an internal GraphQL API
	// this will forward the original request and user info to the internal upstream
	// so that the request context can be enriched
	internal?: boolean;
	skipRenameRootFields?: string[];
	// the schemaExtension field is used to extend the generated GraphQL schema with additional types and fields
	// this is useful for specifying type definitions for JSON objects
	schemaExtension?: string;
	replaceCustomScalarTypeFields?: ReplaceCustomScalarTypeFieldConfiguration[];
}

export interface GraphQLIntrospection extends GraphQLUpstream, GraphQLIntrospectionOptions {
	isFederation?: boolean;
}

export interface GraphQLFederationUpstream extends Omit<Omit<GraphQLUpstream, 'introspection'>, 'apiNamespace'> {
	name?: string;
	loadSchemaFromString?: GraphQLIntrospectionOptions['loadSchemaFromString'];
	introspection?: IntrospectionFetchOptions & GraphqlIntrospectionHeaders;
}

export interface GraphQLFederationIntrospection extends IntrospectionConfiguration {
	upstreams: GraphQLFederationUpstream[];
	apiNamespace?: string;
}

export interface ReplaceCustomScalarTypeFieldConfiguration {
	entityName: string;
	fieldName: string;
	inputTypeReplacement?: string;
	responseTypeReplacement: string;
}

export interface DatabaseIntrospection extends IntrospectionConfiguration {
	databaseURL: InputVariable;
	apiNamespace?: string;
	// the schemaExtension field is used to extend the generated GraphQL schema with additional types and fields
	// this is useful for specifying type definitions for JSON objects
	schemaExtension?: string;
	replaceCustomScalarTypeFields?: ReplaceCustomScalarTypeFieldConfiguration[];
}

export interface PrismaIntrospection extends IntrospectionConfiguration {
	prismaFilePath: string;
	apiNamespace?: string;
	// the schemaExtension field is used to extend the generated GraphQL schema with additional types and fields
	// this is useful for specifying type definitions for JSON objects
	schemaExtension?: string;
	replaceCustomScalarTypeFields?: ReplaceCustomScalarTypeFieldConfiguration[];
}

export interface IntrospectionFetchOptions {
	disableCache?: boolean;
	pollingIntervalSeconds?: number;
}

export interface IntrospectionConfiguration {
	// id is the unique identifier for the data source
	id?: string;
	/**
	 * Timeout for network requests originated by this data source, in seconds.
	 *
	 * @remarks
	 * See {@link NodeOptions| the NodeOptions type} for more details.
	 *
	 * @defaultValue Use the default timeout for this node.
	 */
	requestTimeoutSeconds?: number;
	introspection?: IntrospectionFetchOptions;
}

export interface HTTPUpstream extends IntrospectionConfiguration {
	apiNamespace?: string;
	headers?: (builder: IHeadersBuilder) => IHeadersBuilder;
	authentication?: HTTPUpstreamAuthentication;
	mTLS?: HTTPmTlsConfiguration;
	/** HTTP(S) proxy to use, overriding the default one. To disable a global proxy
	 * set its value to null.
	 *
	 * @defaultValue undefined, which uses the global proxy defined in configureWunderGraphApplication()
	 */
	httpProxyUrl?: InputVariable | null;
}

export type HTTPmTlsConfiguration = {
	/**
	 * 	Private-key or environment variable name that stores the key
	 */
	key: InputVariable;
	/**
	 * 	X.509 (TLS/HTTPS) or environment variable name that stores the certificate
	 */
	cert: InputVariable;
	/**
	 * InsecureSkipVerify controls whether a client verifies the server's certificate chain and host name
	 * If InsecureSkipVerify is true, crypto/tls accepts any certificate presented by the server and any host name in that certificate.
	 * In this mode, TLS is susceptible to machine-in-the-middle attacks unless custom verification is used.
	 * This should be used only for testing
	 */
	insecureSkipVerify: boolean;
};

export type HTTPUpstreamAuthentication = JWTAuthentication | JWTAuthenticationWithAccessTokenExchange;

export interface JWTAuthentication {
	kind: 'jwt';
	secret: InputVariable;
	signingMethod: JWTSigningMethod;
}

export interface JWTAuthenticationWithAccessTokenExchange {
	kind: 'jwt_with_access_token_exchange';
	secret: InputVariable;
	signingMethod: JWTSigningMethod;
	accessTokenExchangeEndpoint: InputVariable;
}

export type JWTSigningMethod = 'HS256';

export interface GraphqlIntrospectionHeaders {
	headers?: (builder: IGraphqlIntrospectionHeadersBuilder) => IGraphqlIntrospectionHeadersBuilder;
}

export interface GraphQLUpstream extends HTTPUpstream {
	url: InputVariable;
	baseUrl?: InputVariable;
	path?: InputVariable;
	subscriptionsURL?: InputVariable;
	subscriptionsUseSSE?: boolean;
	introspection?: IntrospectionFetchOptions & GraphqlIntrospectionHeaders;
}

export interface OpenAPIIntrospectionFile {
	kind: 'file';
	filePath: string;
}

export interface OpenAPIIntrospectionString {
	kind: 'string';
	openAPISpec: string;
}

export interface OpenAPIIntrospectionObject {
	kind: 'object';
	openAPIObject: {};
}

export type OpenAPIIntrospectionSource =
	| OpenAPIIntrospectionFile
	| OpenAPIIntrospectionString
	| OpenAPIIntrospectionObject;

export interface OpenAPIIntrospection extends HTTPUpstream {
	source: OpenAPIIntrospectionSource;
	// statusCodeUnions set to true will make all responses return a union type of all possible response objects,
	// mapped by status code
	// by default, only the status 200 response is mapped, which keeps the GraphQL API flat
	// by enabling statusCodeUnions, you have to unwrap the response union via fragments for each response
	statusCodeUnions?: boolean;
	baseURL?: InputVariable;
	// the schemaExtension field is used to extend the generated GraphQL schema with additional types and fields
	// this is useful for specifying type definitions for JSON objects
	schemaExtension?: string;
	replaceCustomScalarTypeFields?: ReplaceCustomScalarTypeFieldConfiguration[];
}

export interface StaticApiCustom {
	data: ConfigurationVariable;
}

export interface RESTApiCustom {
	Fetch: FetchConfiguration;
	Subscription: SubscriptionConfiguration;
	DefaultTypeName: string;
	StatusCodeTypeMappings: StatusCodeTypeMapping[];
}

export interface DatabaseApiCustom {
	prisma_schema: string;
	graphql_schema: string;
	databaseURL: ConfigurationVariable;
	jsonTypeFields: SingleTypeField[];
	jsonInputVariables: string[];
}

export interface SubscriptionConfiguration {
	Enabled: boolean;
	PollingIntervalMillis?: number;
	SkipPublishSameResponse?: boolean;
}

export interface GraphQLApiCustom {
	Federation: {
		Enabled: boolean;
		ServiceSDL: string;
	};
	Fetch: FetchConfiguration;
	Subscription: {
		Enabled: boolean;
		URL: ConfigurationVariable;
		UseSSE: boolean;
	};
	UpstreamSchema: string;
	HooksConfiguration: GraphQLDataSourceHooksConfiguration;
	CustomScalarTypeFields: SingleTypeField[];
}

export interface GraphQLServerConfiguration extends Omit<GraphQLIntrospection, 'loadSchemaFromString'> {
	schema: GraphQLSchema | Promise<GraphQLSchema>;
}

export const introspectGraphqlServer = async (introspection: GraphQLServerConfiguration) => {
	const { schema, ...rest } = introspection;
	const resolvedSchema = (await schema) as GraphQLSchema;

	return introspectGraphqlWithCache({
		...rest,
		internal: true,
		loadSchemaFromString: () => printSchema(buildClientSchema(introspectionFromSchema(resolvedSchema))),
	});
};

export const introspect = {
	graphql: (introspection: Omit<GraphQLIntrospection, 'isFederation'>) => {
		return introspectGraphqlWithCache(introspection);
	},
	postgresql: introspectPostgresql,
	mysql: introspectMySQL,
	planetscale: introspectPlanetScale,
	sqlite: introspectSQLite,
	sqlserver: introspectSQLServer,
	mongodb: introspectMongoDB,
	prisma: introspectPrisma,
	federation: introspectFederation,
	openApi: introspectOpenApi,
	openApiV2: introspectOpenApiV2,
	soap: introspectSoap,
};

export const buildUpstreamAuthentication = (upstream: HTTPUpstream): UpstreamAuthentication | undefined => {
	if (upstream.authentication === undefined) {
		return undefined;
	}
	return {
		kind: upstreamAuthenticationKind(upstream.authentication.kind),
		jwtConfig:
			upstream.authentication.kind === 'jwt'
				? {
						secret: mapInputVariable(upstream.authentication.secret),
						signingMethod: upstreamAuthenticationSigningMethod(upstream.authentication.signingMethod),
				  }
				: undefined,
		jwtWithAccessTokenExchangeConfig:
			upstream.authentication.kind === 'jwt_with_access_token_exchange'
				? {
						accessTokenExchangeEndpoint: mapInputVariable(upstream.authentication.accessTokenExchangeEndpoint),
						secret: mapInputVariable(upstream.authentication.secret),
						signingMethod: upstreamAuthenticationSigningMethod(upstream.authentication.signingMethod),
				  }
				: undefined,
	};
};

export const buildMTLSConfiguration = (upstream: HTTPUpstream): MTLSConfiguration | undefined => {
	if (upstream.mTLS === undefined) {
		return undefined;
	}
	return {
		key: mapInputVariable(upstream.mTLS?.key || ''),
		cert: mapInputVariable(upstream.mTLS?.cert || ''),
		insecureSkipVerify: upstream.mTLS?.insecureSkipVerify || false,
	};
};

const upstreamAuthenticationSigningMethod = (signingMethod: JWTSigningMethod): SigningMethod => {
	switch (signingMethod) {
		case 'HS256':
			return SigningMethod.SigningMethodHS256;
		default:
			throw new Error(`JWT signing method unsupported: ${signingMethod}`);
	}
};

const upstreamAuthenticationKind = (kind: HTTPUpstreamAuthentication['kind']): UpstreamAuthenticationKind => {
	switch (kind) {
		case 'jwt':
			return UpstreamAuthenticationKind.UpstreamAuthenticationJWT;
		case 'jwt_with_access_token_exchange':
			return UpstreamAuthenticationKind.UpstreamAuthenticationJWTWithAccessTokenExchange;
		default:
			throw new Error(`upstreamAuthenticationKind, unsupported kind: ${kind}`);
	}
};
