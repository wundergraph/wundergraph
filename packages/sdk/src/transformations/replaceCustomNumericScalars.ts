import { GraphQLIntrospection } from '../definition';
import { parse, visit } from 'graphql';
import { printSchemaWithDirectives } from '@graphql-tools/utils';
import { buildASTSchema } from 'graphql/index';

export interface ArgumentReplacement {
	fieldName: string;
	typeName: string;
	argName: string;
	renameTypeTo: string;
}

export interface ReplaceCustomNumericScalarsResult {
	schemaSDL: string;
	argumentReplacements: ArgumentReplacement[];
}

export const replaceCustomNumericScalars = (
	schemaSDL: string,
	introspection: GraphQLIntrospection
): ReplaceCustomNumericScalarsResult => {
	const customFloatScalars = introspection.customFloatScalars || [];
	const customIntScalars = introspection.customIntScalars || [];
	const argumentReplacements: ArgumentReplacement[] = [];

	let typeName: undefined | string;
	let fieldName: undefined | string;
	let argName: undefined | string;

	const ast = parse(schemaSDL);
	const cleanAst = visit(ast, {
		InterfaceTypeDefinition: {
			enter: (node) => {
				typeName = node.name.value;
			},
			leave: () => {
				typeName = undefined;
			},
		},
		ObjectTypeExtension: {
			enter: (node) => {
				typeName = node.name.value;
			},
			leave: () => {
				typeName = undefined;
			},
		},
		InterfaceTypeExtension: {
			enter: (node) => {
				typeName = node.name.value;
			},
			leave: () => {
				typeName = undefined;
			},
		},
		ObjectTypeDefinition: {
			enter: (node) => {
				typeName = node.name.value;
			},
			leave: (node) => {
				typeName = undefined;
			},
		},
		FieldDefinition: {
			enter: (node) => {
				fieldName = node.name.value;
			},
			leave: (node) => {
				fieldName = undefined;
			},
		},
		InputValueDefinition: {
			enter: (node) => {
				argName = node.name.value;
			},
			leave: (node) => {
				argName = undefined;
			},
		},
		ScalarTypeDefinition: (node) => {
			if (customFloatScalars.includes(node.name.value)) {
				return null;
			}
			if (customIntScalars.includes(node.name.value)) {
				return null;
			}
		},
		NamedType: (node) => {
			if (customFloatScalars.includes(node.name.value)) {
				if (argName && fieldName && typeName) {
					argumentReplacements.push({
						argName,
						fieldName,
						typeName,
						renameTypeTo: node.name.value,
					});
				}
				return {
					...node,
					name: {
						...node.name,
						value: 'Float',
					},
				};
			}
			if (customIntScalars.includes(node.name.value)) {
				if (argName && fieldName && typeName) {
					argumentReplacements.push({
						argName,
						fieldName,
						typeName,
						renameTypeTo: node.name.value,
					});
				}
				return {
					...node,
					name: {
						...node.name,
						value: 'Int',
					},
				};
			}

			return node;
		},
	});

	const schema = printSchemaWithDirectives(buildASTSchema(cleanAst));
	return { schemaSDL: schema, argumentReplacements };
};
