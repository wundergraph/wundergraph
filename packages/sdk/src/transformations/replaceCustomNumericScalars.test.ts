import { GraphQLIntrospection } from '../definition';
import { describe } from 'node:test';
import { replaceCustomNumericScalars } from './replaceCustomNumericScalars';

interface TestCase {
	schema: string;
	introspection: GraphQLIntrospection;
}

const runTest = (testCase: TestCase, snapshotName: string) => {
	const result = replaceCustomNumericScalars(testCase.schema, testCase.introspection);
	expect(result).toMatchSnapshot(snapshotName);
};

describe('replaceCustomNumericScalars', () => {
	it('should replace int scalar', function () {
		const testCase = {
			schema: `
					scalar CustomInt
					type Query {
						foo: CustomInt
						bar(a: CustomInt): String
					}
				`,
			introspection: {
				url: 'http://dummy',
				customIntScalars: ['CustomInt'],
			},
		};

		runTest(testCase, 'custom int scalar');
	});

	it('should replace float scalar', function () {
		const testCase = {
			schema: `
					scalar CustomFloat
					type Query {
						foo: CustomFloat
						bar(a: CustomFloat): String
					}
					`,
			introspection: {
				url: 'http://dummy',
				customFloatScalars: ['CustomFloat'],
			},
		};

		runTest(testCase, 'custom float scalar');
	});
});
