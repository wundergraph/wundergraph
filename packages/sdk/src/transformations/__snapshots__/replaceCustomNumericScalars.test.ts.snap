// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`replaceCustomNumericScalars should replace float scalar: custom float scalar 1`] = `
{
  "argumentReplacements": [
    {
      "argName": "a",
      "fieldName": "bar",
      "renameTypeTo": "CustomFloat",
      "typeName": "Query",
    },
  ],
  "schemaSDL": "type Query {
  foo: Float
  bar(a: Float): String
}",
}
`;

exports[`replaceCustomNumericScalars should replace int scalar: custom int scalar 1`] = `
{
  "argumentReplacements": [
    {
      "argName": "a",
      "fieldName": "bar",
      "renameTypeTo": "CustomInt",
      "typeName": "Query",
    },
  ],
  "schemaSDL": "type Query {
  foo: Int
  bar(a: Int): String
}",
}
`;
